### PR TITLE
Removed allocation in object Tag

### PR DESCRIPTION
**Pull request by Arneball**

### CODE REVIEW

**Comment by milessabin** [comment]
This is fine, but it's binary incompatible so it'll have to wait until the 2.4.0.

Can you point to an example project where the allocation you're removing here is an issue in practice?

**Comment by Atry** [comment]
IIRC, Hotspot JVM can easily eliminate these kind of allocation.

**Comment by Arneball** [comment]
Dalvik cannot do escape analysis AFAIK. Saw the allocation in a toy project
of mine.

Den fre 21 apr. 2017 15:21杨博 (Yang Bo) <notifications@github.com> skrev:

File: core/src/main/scala/shapeless/typeoperators.scala [hunk 1]
```diff
@@ -23,14 +23,15 @@ import scala.reflect.macros.whitebox
...
 
 object tag {
-  def apply[U] = new Tagger[U]
+  def apply[U] = Tagger.asInstanceOf[Tagger[U]]
 
   trait Tagged[U]
   type @@[+T, U] = T with Tagged[U]
 
-  class Tagger[U] {
+  trait Tagger[U] {
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/typeoperators.scala]: Why does it have to be a trait?

Comment (joroKr21) [reply core/src/main/scala/shapeless/typeoperators.scala]:
```suggestion
  class Tagger[U] {
```

**Pull request merged**